File: frontend/src/types/employer-outreach.ts
export type CandidateWorkflowNote = {
  id: string;
  author: string;
  body: string;
  createdAt: string;
  tone?: 'neutral' | 'success' | 'warning' | 'ai';
};

export type CandidateWorkflowState = {
  tags: string[];
  notes: CandidateWorkflowNote[];
};

export type CandidateWorkflowById = Record<number, CandidateWorkflowState>;

export type EmployerOutreachChannel = 'email' | 'linkedin' | 'whatsapp';
export type EmployerOutreachTone = 'warm' | 'direct' | 'premium' | 'technical';
export type EmployerOutreachTemplate = 'role-fit' | 'quick-intro' | 'technical-depth' | 'salary-transparent';

export type EmployerOutreachCampaign = {
  id: string;
  name: string;
  status: 'draft' | 'queued' | 'sent' | string;
  candidate_ids: number[];
  channel: EmployerOutreachChannel;
  tone: EmployerOutreachTone;
  template: EmployerOutreachTemplate;
  include_salary: boolean;
  include_calendar: boolean;
  message_preview: string;
  response_rate: number;
  created_at: string;
  updated_at?: string;
  metadata?: Record<string, unknown>;
};

export type EmployerOutreachCampaignCreatePayload = {
  name?: string;
  candidate_ids: number[];
  channel: EmployerOutreachChannel;
  tone: EmployerOutreachTone;
  template: EmployerOutreachTemplate;
  include_salary: boolean;
  include_calendar: boolean;
  message_preview?: string;
  response_rate: number;
  metadata?: Record<string, unknown>;
};

export type EmployerSendQueueItemStatus = 'ready' | 'review' | 'blocked';

export type EmployerSendQueueAuditEvent = {
  id: string;
  action: string;
  actor: string;
  note: string;
  created_at: string;
  metadata?: Record<string, unknown>;
};

export type EmployerSendQueueItem = {
  id: string;
  candidate_id: number;
  subject: string;
  message: string;
  response_score: number;
  status: EmployerSendQueueItemStatus;
  checks: string[];
  metadata?: Record<string, unknown>;
  updated_at?: string;
};

export type EmployerSendQueue = {
  id: string;
  campaign_id: string;
  employer_id?: string;
  status: 'ready_for_review' | 'blocked' | string;
  channel?: EmployerOutreachChannel;
  items: EmployerSendQueueItem[];
  ready_count: number;
  review_count: number;
  blocked_count: number;
  created_at: string;
  updated_at: string;
  audit: EmployerSendQueueAuditEvent[];
};

export type EmployerSendQueuePreparePayload = {
  items: Array<{
    candidate_id: number;
    subject: string;
    message: string;
    response_score: number;
    status: EmployerSendQueueItemStatus;
    checks: string[];
    metadata?: Record<string, unknown>;
  }>;
  actor?: string;
  note?: string;
};

export type EmployerSendQueueItemUpdatePayload = {
  status?: EmployerSendQueueItemStatus;
  checks?: string[];
  actor?: string;
  note?: string;
};

File: src/hiring_radar/api/schemas/employer_outreach.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

OutreachChannel = Literal["email", "linkedin", "whatsapp"]
OutreachTone = Literal["warm", "direct", "premium", "technical"]
OutreachTemplate = Literal["role-fit", "quick-intro", "technical-depth", "salary-transparent"]


class EmployerOutreachCampaignCreateRequest(BaseModel):
    name: str | None = None
    candidate_ids: list[int] = Field(default_factory=list, min_length=1)
    channel: OutreachChannel
    tone: OutreachTone
    template: OutreachTemplate
    include_salary: bool = True
    include_calendar: bool = True
    message_preview: str | None = None
    response_rate: int = Field(default=0, ge=0, le=100)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EmployerCandidateBulkTagRequest(BaseModel):
    candidate_ids: list[int] = Field(default_factory=list, min_length=1)
    tag: str = Field(min_length=1, max_length=64)


class EmployerCandidateBulkNoteRequest(BaseModel):
    candidate_ids: list[int] = Field(default_factory=list, min_length=1)
    note: str = Field(min_length=1, max_length=2000)
    tone: Literal["neutral", "success", "warning", "ai"] = "neutral"

QueueItemStatus = Literal["ready", "review", "blocked"]


class EmployerSendQueueItemCreateRequest(BaseModel):
    candidate_id: int
    subject: str = Field(min_length=1, max_length=240)
    message: str = Field(min_length=1, max_length=6000)
    response_score: int = Field(default=0, ge=0, le=100)
    status: QueueItemStatus = "review"
    checks: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EmployerSendQueuePrepareRequest(BaseModel):
    items: list[EmployerSendQueueItemCreateRequest] = Field(default_factory=list, min_length=1)
    actor: str = Field(default="Sen", min_length=1, max_length=120)
    note: str | None = Field(default=None, max_length=1000)


class EmployerSendQueueItemUpdateRequest(BaseModel):
    status: QueueItemStatus | None = None
    checks: list[str] | None = None
    actor: str = Field(default="Sen", min_length=1, max_length=120)
    note: str | None = Field(default=None, max_length=1000)


class EmployerCandidateStageUpdateRequest(BaseModel):
    candidate_id: int = Field(ge=1)
    stage_key: Literal["new", "reviewed", "shortlist", "interview", "offer", "hired", "archived"]
    job_id: int | None = Field(default=None, ge=1)
    note: str | None = Field(default=None, max_length=1000)
